crates/torii/grpc/src/server/subscriptions/entity.rs

```diff
@@ -88,64 +87,106 @@ impl Service {
         subs: Arc<EntityManager>,
         cache: Arc<ModelCache>,
         pool: Pool<Sqlite>,
-        hashed_keys: &str,
+        entity: &Entity,
     ) -> Result<(), Error> {
         let mut closed_stream = Vec::new();
+        let hashed = FieldElement::from_str(&entity.id).map_err(ParseError::FromStr)?;
+        let keys = entity
+            .keys
+            .trim_end_matches(FELT_DELIMITER)
+            .split(FELT_DELIMITER)
+            .map(FieldElement::from_str)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(ParseError::FromStr)?;
 
         for (idx, sub) in subs.subscribers.read().await.iter() {
-            let hashed = FieldElement::from_str(hashed_keys).map_err(ParseError::FromStr)?;
-            // publish all updates if ids is empty or only ids that are subscribed to
-            if sub.hashed_keys.is_empty() || sub.hashed_keys.contains(&hashed) {
-                let models_query = r#"
+            // Check if the subscriber is interested in this entity
+            // If we have a clause of hashed keys, then check that the id of the entity
+            // is in the list of hashed keys.
+
+            // If we have a clause of keys, then check that the key pattern of the entity
+            // matches the key pattern of the subscriber.
+            match &sub.keys {
+                Some(EntityKeysClause::HashedKeys(hashed_keys)) => {
+                    if !hashed_keys.contains(&hashed) {
+                        continue;
+                    }
+                }
+                Some(EntityKeysClause::Keys(clause)) => {
+                    // if the key pattern doesnt match our subscribers key pattern, skip
+                    // ["", "0x0"] would match with keys ["0x...", "0x0", ...]
+                    if !keys.iter().enumerate().all(|(idx, key)| {
```

Comment:
As mentioned in the call, will put this here as a reminder:
we should check if the key array is smaller, to ensure we don't match a shorter array of keys in fixed len. 👍